File: docking/docking/drone.py
# standard python library imports
import asyncio
import time
import math
import numpy as np

# mavsdk
from mavsdk import System
from mavsdk.offboard import (OffboardError, VelocityBodyYawspeed, VelocityNedYaw)

# aviata modules
import util
from camera_simulator import CameraSimulator
from image_analyzer import ImageAnalyzer
from debug_window import DebugWindow
from pid_controller import PIDController
from log import Log

boom_length = 1 # meter
fov = 48.8
altitude_disp = boom_length / 2 / math.tan(math.radians(fov / 2)) * 2 # add 100% tolerance

class Drone:

    def __init__(self, target):
        self.camera_simulator = CameraSimulator(target) # later, use camera here
        self.image_analyzer = ImageAnalyzer()
        self.north = 0
        self.east = 0
        self.down = -8 # initial altitude
        self.yaw = 0
        self.target = target 
        self.dt = 0.05 # delta t update frequency, 20hz
        self.MAX_ATTEMPTS = 3 # number of unsuccessful attempts before aborting docking
        self.MAX_HEIGHT = 10 # max height above central target before failure
        self.MAX_HEIGHT_STAGE_2 = 2
        self.STAGE_1_TOLERANCE = 0.1 
        self.STAGE_2_TOLERANCE = 0.05
        self.log=Log()
        self.logging=True

    async def connect_gazebo(self):
        """
        Connect to gazebo and perform basic setup checks.
        Does not return until connection to drone is verified.
        """
        # Wait for connection
        self.drone = System()
        await self.drone.connect(system_address="udp://:14540")

        async for state in self.drone.core.connection_state():
            if state.is_connected:
                print(f"Connected to drone with UUID: {state.uuid}")
                break

        # Basic setup checks
        async for global_lock in self.drone.telemetry.health():
            if global_lock.is_global_position_ok:
                print("-- Global position state is ok")
                break

    async def takeoff(self):
        """
        Takeoff and start offboard mode.
        Return true if success, false otherwise
        """
        print("-- Arming")
        await self.drone.action.arm()

        # Takeoff the vehicle (later, we can assume that the drone is already flying when we run this script)
        print("-- Taking off")
        await self.drone.action.set_takeoff_altitude(3.0) # meters
        await self.drone.action.takeoff()
        await asyncio.sleep(8)

        # start offboard mode (requires setting initial setpoint)
        print("-- Setting initial setpoint")
        await self.drone.offboard.set_velocity_ned(
            VelocityNedYaw(0.0, 0.0, 0.0, 135))

        print("-- Starting offboard")
        try:
            await self.drone.offboard.start()
        except OffboardError as error:
            print(f"Starting offboard mode failed with error code: \
                {error._result.result}")
            print("-- Disarming")
            await self.drone.action.disarm()
            return False

        return True

    async def initiate_docking(self, id):

        # Get telemetry periodically from the drone
        telemetry_task = asyncio.create_task(self.get_telemetry_position())
        rotation_task = asyncio.create_task(self.get_telemetry_rotation())

        if await self.stage1(id):
            await self.stage2(id)
            await self.land()
        else:
            print("Docking failed: Max height exceeded")
            await self.safe_land()

        telemetry_task.cancel()
        rotation_task.cancel()

    async def stage1(self, id):
        """Position the drone above the large central target, returns true if successful"""
        print("Docking stage 1")
        self.log.writeLiteral("Docking stage 1")

        debug_window=DebugWindow(1,self.target)

        failed_frames = 0 # Number of frames since we detected the target
        successful_frames = 0 # Number of frames within tolerance 
        pid_controller = PIDController(self.dt)

        while True:
            start_millis = int(round(time.time() * 1000))
            img = self.camera_simulator.updateCurrentImage(self.east, self.north, self.down * -1.0, self.yaw)
            errs = self.image_analyzer.process_image(img, 0, self.yaw)
            
            if errs is None:
                self.log.writeLiteral("No Errors")
                failed_frames = failed_frames + 1

                if self.down * -1.0 - self.target.getAlt() > self.MAX_HEIGHT: # drone moves above max height, docking fails
                    return False

                if failed_frames > 1 / self.dt: # haven't detected target in 1 second
                    await self.drone.offboard.set_velocity_body(
                        VelocityBodyYawspeed(0, 0, -0.2, 0))

                await asyncio.sleep(self.dt)
                continue

            failed_frames = 0
            x_err, y_err, alt_err, rot_err, tags_detected = errs

            debug_window.updateWindow(self.east, self.north, self.down * -1.0, self.yaw, tags_detected)
            if self.logging:
                self.log.write(self.east, self.north, self.down * -1.0, self.yaw, tags_detected,successful_frames,x_err,y_err,alt_err,rot_err)

            x_err, y_err, alt_err = self.offset_errors(x_err, y_err, alt_err, rot_err, id)
            east_velocity, north_velocity, down_velocity = pid_controller.get_velocities(x_err, y_err, alt_err, 0.4)
            
            # Checks if drone is aligned with central target
            if util.is_between_symm(alt_err, self.STAGE_1_TOLERANCE) and util.is_between_symm(x_err, self.STAGE_1_TOLERANCE) and util.is_between_symm(y_err, self.STAGE_1_TOLERANCE):
                successful_frames += 1
            else:
                successful_frames = 0

            # Ends loop if aligned for 1 second
            if successful_frames == 1 / self.dt:
                break

            await self.drone.offboard.set_velocity_ned(
                VelocityNedYaw(north_m_s=north_velocity, east_m_s=east_velocity, down_m_s=down_velocity, yaw_deg=rot_err)) # north, east, down (all m/s), yaw (degrees, north is 0, positive for clockwise)

            current_millis = int(round(time.time() * 1000))
            if current_millis - start_millis < self.dt:
                await asyncio.sleep(self.dt - (current_millis - start_millis))
        
        debug_window.destroyWindow()
        return True

    async def stage2(self, id):
        """Position the drone above the peripheral target and descend"""
        print("Docking stage 2")
        self.log.writeLiteral("Docking stage 2")

        debug_window = DebugWindow(2,self.target)
        pid_controller = PIDController(self.dt)
        successful_frames = 0
        while True:
        
            img = self.camera_simulator.updateCurrentImage(self.east, self.north, self.down * -1.0, self.yaw,id)
            errs = self.image_analyzer.process_image(img, id, self.yaw)
            checked_frames = 0
            docking_attempts = 0

            # Waits for one second to detect target tag, ascends to find central target if fails
            while errs is None:
                checked_frames += 1
                self.log.writeLiteral("No errors, checked frames: "+str(checked_frames))
                await self.drone.offboard.set_velocity_body(VelocityBodyYawspeed(0, 0, -0.1, 0))

                if checked_frames > 1 / self.dt:
                    docking_attempts += 1
                    if docking_attempts > self.MAX_ATTEMPTS:
                        print("Docking failed")
                        self.log.writeLiteral("Docking failed")
                        await self.safe_land()
                        return

                    img = self.camera_simulator.updateCurrentImage(self.east, self.north, self.down * -1.0, self.yaw,id)
                    errs = self.image_analyzer.process_image(img,0,self.yaw)
                    
                    # Ascends until maximum height or until peripheral target detected
                    while errs is None and self.down * -1.0 - self.target.getAlt() < self.MAX_HEIGHT_STAGE_2:
                        await self.drone.offboard.set_velocity_body(VelocityBodyYawspeed(0, 0, -0.2, 0))
                        img = self.camera_simulator.updateCurrentImage(self.east, self.north, self.down * -1.0, self.yaw, id)
                        errs = self.image_analyzer.process_image(img, id, self.yaw)

                    if not errs is None:
                        break

                    # If peripheral tag not found, ascends until maximum height or until central target detected
                    while errs is None and self.down * -1.0 - self.target.getAlt() < self.MAX_HEIGHT:
                        await self.drone.offboard.set_velocity_body(VelocityBodyYawspeed(0, 0, -0.2, 0))
                        img = self.camera_simulator.updateCurrentImage(self.east, self.north, self.down * -1.0, self.yaw)
                        errs = self.image_analyzer.process_image(img,0, self.yaw)
                    
                    # Re-attempts stage 1
                    if not errs is None:
                        await self.stage1(id)
                        img = self.camera_simulator.updateCurrentImage(self.east, self.north, self.down * -1.0, self.yaw)
                        errs = self.image_analyzer.process_image(img, id, self.yaw)
                        checked_frames = 0
                    else: # If the central target cannot be found at maximum height (maybe could have re-attempt stage 1 again, not sure)
                        print("Docking failed")
                        await self.safe_land()
                        return 
                        

                await asyncio.sleep(self.dt)
                img = self.camera_simulator.updateCurrentImage(self.east, self.north, self.down * -1.0, self.yaw,id)
                errs = self.image_analyzer.process_image(img, id,self.yaw)

            x_err, y_err, alt_err, rot_err, tags_detected = errs
            alt_err = alt_err - .05 
            if self.logging:
                self.log.write(self.east, self.north, self.down * -1.0, self.yaw, tags_detected,successful_frames,x_err,y_err,alt_err,rot_err)

            # Checks if drone is aligned with docking
            if alt_err < self.STAGE_2_TOLERANCE * 2 and alt_err > -1 * self.STAGE_2_TOLERANCE and rot_err < 2.0 and rot_err > -2.0 and x_err > -1 * self.STAGE_2_TOLERANCE and x_err < 1 * self.STAGE_2_TOLERANCE and y_err > -1 * self.STAGE_2_TOLERANCE and y_err<self.STAGE_2_TOLERANCE:
                successful_frames += 1
            else:
                successful_frames = 0

            # Adjusts errors for distance to target to prevent overshoot
            # Adjusted errors asymptote to 1 as alt_err increases, goes to 0 as alt_err decreases
            OVERSHOOT_CONSTANT=.6 # use to adjust speed of descent, higher constant means faster, lower means less overshooting
            # x_err*=np.tanh(OVERSHOOT_CONSTANT*alt_err*alt_err)
            # y_err*=np.tanh(OVERSHOOT_CONSTANT*alt_err*alt_err)
            # rot_err*=np.tanh(OVERSHOOT_CONSTANT*alt_err*alt_err)
            # alt_err*=np.tanh(OVERSHOOT_CONSTANT*alt_err*alt_err)

            # Ends loop if aligned for 1 second
            if successful_frames == 1 / self.dt:
                break

            debug_window.updateWindow(self.east, self.north, self.down * -1.0, self.yaw, tags_detected)

            east_velocity, north_velocity, down_velocity = pid_controller.get_velocities(x_err, y_err, alt_err, 0.1)

            await self.drone.offboard.set_velocity_ned(
                VelocityNedYaw(north_m_s=north_velocity, east_m_s=east_velocity, down_m_s=down_velocity, yaw_deg=rot_err)) # north, east, down (all m/s), yaw (degrees, north is 0, positive for clockwise)
            
            await asyncio.sleep(self.dt)

    def offset_errors(self, x, y, alt, rot, target):
        # maps target 3 to 0 degrees, target 2 to 45 degrees, 1 to 90, 8 to 135, 7 to 180, ... , 4 to 315
        # note that peripheral targets are numbered so that the north one is 1, then increasing clockwise
        target_offset = abs(target - 3) * 45 if target <= 3 else (11 - target) * 45

        x = x + boom_length / 2 * math.cos(math.radians(target_offset - rot))
        y = y + boom_length / 2 * math.sin(math.radians(target_offset - rot))
        
        alt = alt - altitude_disp
        return x, y, alt
    
    # Moves drone up and away from current location, then ends offboard control and lands
    async def safe_land(self):
        dt=0.05
        for i in range (5/0.05): #Drone flies away up and to the north for 5 seconds 
            await self.drone.offboard.set_velocity_ned(
                        VelocityNedYaw(0.4, 0, -0.4, 0.0)) #These numbers can be changed to get out of GPS error margin
            asyncio.sleep(dt)
        
        await self.land()

    async def get_telemetry_rotation(self):
        """Poll to obtain the drone's yaw"""
        await self.drone.telemetry.set_rate_attitude(20) # update at 20Hz 
        async for angles in self.drone.telemetry.attitude_euler():
            self.yaw = angles.yaw_deg # 0 degrees is north, positive cw, negative ccw, between -180 and +180


    async def get_telemetry_position(self):
        """Poll to obtain the drone's x, y, and z position"""
        await self.drone.telemetry.set_rate_position_velocity_ned(20) # update at 20Hz 
        async for position in self.drone.telemetry.position_velocity_ned():
            p = position.position
            self.north = p.north_m
            self.east = p.east_m
            self.down = p.down_m

    async def land(self):
        """Safely stop the drone and stop offboard mode"""
        print("-- Stopping offboard")
        try:
            await self.drone.offboard.stop()
        except OffboardError as error:
            print(f"Stopping offboard mode failed with error code: \
                {error._result.result}")

        print("-- Landing drone")
        await self.drone.action.land()

File: docking/cpp_docking/src/drone.cpp
#include "drone.hpp"
#include "util.hpp"
#include "pid_controller.hpp"

#include <iostream>
#include <future>
#include <string>
#include <memory>

#include <opencv2/core.hpp>
#include <opencv2/core/types.hpp>
#include <opencv2/imgcodecs.hpp>

#include <mavsdk/mavsdk.h>
#include <mavsdk/plugins/action/action.h>
#include <mavsdk/plugins/offboard/offboard.h>
#include <mavsdk/plugins/telemetry/telemetry.h>

using namespace mavsdk;
using namespace std::chrono;
using std::this_thread::sleep_for;

Drone::Drone(Target t)
    : camera(t), image_analyzer(), m_target_info(t), m_north(0), m_east(0), m_down(-5), m_yaw(0)
{
}

/**
 * Connect to Gazebo and run pre-flight checks
 * This function should be called before attempting to take any other action with the drone.
 * 
 * @return true if connection succeeded, false otherwise
 * */
bool Drone::connect_px4()
{
    std::string tag = "Connecting";

    std::string connection_url;
#if PLATFORM == RASPBERRY_PI
    connection_url = "serial:///dev/ttyAMA0:921600";
#else
    connection_url = "udp://:14540";
#endif

    ConnectionResult connection_result = mavsdk.add_any_connection(connection_url);

    if (connection_result != ConnectionResult::Success)
    {
        log(tag, "Connection failed, see next line", true);
        std::cout << connection_result << std::endl;
        return false;
    }

    log(tag, "Waiting to discover system");
    std::promise<void> discover_promise;
    auto discover_future = discover_promise.get_future();

    mavsdk.subscribe_on_new_system([&mavsdk = mavsdk, &discover_promise, &tag]()
                                   {
                                       const auto system = mavsdk.systems().at(0);

                                       if (system->is_connected())
                                       {
                                           log(tag, "Discovered system");
                                           discover_promise.set_value();
                                       }
                                   });

    discover_future.wait();

    // System got discovered.
    m_system = mavsdk.systems().at(0);
    telemetry = std::make_shared<Telemetry>(m_system);
    action = std::make_shared<Action>(m_system);
    offboard = std::make_shared<Offboard>(m_system);

    while (!telemetry->health_all_ok())
    {
        log(tag, "Waiting for system to be ready");
        // log(tag, "is_home_position_ok " + std::to_string(telemetry->health().is_home_position_ok));
        // log(tag, "is_global_position_ok " + std::to_string(telemetry->health().is_global_position_ok));
        // log(tag, "is_local_position_ok " + std::to_string(telemetry->health().is_local_position_ok));
        // log(tag, "is_level_calibration_ok " + std::to_string(telemetry->health().is_level_calibration_ok));
        // log(tag, "is_magnetometer_calibration_ok " + std::to_string(telemetry->health().is_magnetometer_calibration_ok));
        // log(tag, "is_accelerometer_calibration_ok " + std::to_string(telemetry->health().is_accelerometer_calibration_ok));
        // log(tag, "is_gyrometer_calibration_ok " + std::to_string(telemetry->health().is_gyrometer_calibration_ok));
        sleep_for(seconds(1));
    }

    log(tag, "System is ready");
    return true;
}

/**
 * Attempt to arm the drone.
 * 
 * @return true if successful, false otherwise.
 * */
bool Drone::arm()
{
    std::string tag = "Arming";

    // Arm the drone
    Action::Result arm_result = action->arm();
    if (arm_result != Action::Result::Success)
    {
        log(tag, "Arming failed", true);
        return false;
    }
    log(tag, "Armed");
    return true;
}

/**
 * Attempt to take off and start offboard mode. Drone must be armed first.
 * 
 * @return true if successful, false otherwise
 * */
bool Drone::takeoff(int takeoff_alt)
{
    std::string tag = "Takeoff";
    std::promise<void> in_air_promise;
    auto in_air_future = in_air_promise.get_future();

    // Attempt to take off
    action->set_takeoff_altitude(takeoff_alt);
    Action::Result takeoff_result = action->takeoff();
    if (takeoff_result != Action::Result::Success)
    {
        log(tag, "Takeoff failed, error code on next line", true);
        std::cout << "Code: " << takeoff_result << std::endl;
        return false;
    }

    log(tag, "Taking off");

    // Wait for the take off to be complete
    telemetry->subscribe_landed_state([&in_air_promise, this, &tag](Telemetry::LandedState landed)
                                      {
                                          std::cout << "LandedState: " << landed << std::endl;
                                          if (landed == Telemetry::LandedState::InAir)
                                          {
                                              telemetry->subscribe_landed_state(nullptr);
                                              in_air_promise.set_value();
                                              log(tag, "Take off successful");
                                          }
                                      });

    in_air_future.wait(); // wait for the asynchronous call to be finished (when in_air_promise resolves)

    // Start offboard mode
    Offboard::VelocityBodyYawspeed stay{};
    offboard->set_velocity_body(stay); // set initial setpoint

    Offboard::Result offboard_result = offboard->start();
    if (offboard_result != Offboard::Result::Success)
    {
        log(tag, "Offboard start failed, error code on next line", true);
        std::cout << "Code: " << offboard_result << std::endl;
        return false;
    }

    log(tag, "Offboard started");

    return true;
}

void Drone::set_position(float north, float east, float down)
{
    m_north = north;
    m_east = east;
    m_down = down;
}

void Drone::set_yaw(float yaw)
{
    m_yaw = yaw;
}

void Drone::warm_camera()
{
    log("Drone Pre-flight", "Warming camera up...");
    for (int count = 0; count < 60; count++)
    {
        camera.update_current_image(0, 0, 5, 0, 0);
    }
}

void Drone::initiate_docking(int target_id, bool full_docking)
{
    // set up telemetry updates; happens in background through callbacks
    const Telemetry::Result set_rate_result_p = telemetry->set_rate_position_velocity_ned(20);
    const Telemetry::Result set_rate_result_a = telemetry->set_rate_attitude(20);
    if (set_rate_result_a != Telemetry::Result::Success || set_rate_result_p != Telemetry::Result::Success)
    {
        std::cout << "Setting attitude rate possibly failed:" << set_rate_result_a << '\n';
        std::cout << "Setting position rate possibly failed:" << set_rate_result_p << '\n';
    }

    // Subscribe to getting periodic telemetry updates with lambda function
    telemetry->subscribe_position_velocity_ned([this](Telemetry::PositionVelocityNed p)
                                               { set_position(p.position.north_m, p.position.east_m, p.position.down_m); });
    telemetry->subscribe_attitude_euler([this](Telemetry::EulerAngle e)
                                        { set_yaw(e.yaw_deg); });

    //Disengage servo if not already
    #if PLATFORM == RASPBERRY_PI
        docking_detector.disengage_servo();
#endif

    // begin docking
    if (full_docking)
    { // do complete 2-stage docking
        bool stage1_success = dock(target_id, STAGE_1);
        if (stage1_success)
        {
            dock(target_id, STAGE_2);
        }
    }
    else
    {                             // do 1-stage docking (used for testing)
        dock(target_id, STAGE_2); // tag id = 0, central target
    }

    telemetry->subscribe_position_velocity_ned(nullptr);
    telemetry->subscribe_attitude_euler(nullptr);

    land();
}

/**
 * Attempts a particular stage of the docking process
 * Previously used to have 1 function per stage, but code got too long/duplicated
 * Easier to have 1 function with conditional separation of logic for small parts that are unique to a stage
 * 
 * @param target_id is the target to dock to
 * @param stage is which stage we're in
 * @return true if successful
 * */
bool Drone::dock(int target_id, int stage)
{
    std::string log_tag = "Stage " + std::to_string(stage);
    log(log_tag, "Docking Beginning");

    // Initialize all persistent variables
    PIDController pid;
    Mat img;
    int failed_frames = 0;
    int successful_frames = 0;
    int time_span = 0;
    int total_frames = 0;
    Velocities velocities;
    high_resolution_clock::time_point t1 = high_resolution_clock::now();
    bool success = false;

    while (time_span < 60 * 1000)
    {
        Errors errs;
        Offboard::VelocityBodyYawspeed change{};
        std::string tags_detected = "";

        img = camera.update_current_image(m_east, m_north, m_down * -1.0, m_yaw, 0);
        bool is_tag_detected = image_analyzer.processImage(img, stage == STAGE_1 ? 0 : target_id, tags_detected, errs); // central target has id = 0

        if (is_tag_detected)
        {
            failed_frames = 0;
            if (stage == STAGE_1)
                offset_errors(errs, target_id); // adjusts errors for stage 1 to stage 2 transition
            else
                errs.alt -= 0.35; //0.30;
            velocities = pid.getVelocities(errs.x, errs.y, errs.alt, errs.yaw, 1.5);
            log(log_tag, "Apriltag found! Timestamp: " + std::to_string(time_span) +
                             " errors: x=" + std::to_string(errs.x) + " y=" + std::to_string(errs.y) + " z=" + std::to_string(errs.alt) + " yaw=" + std::to_string(errs.yaw) + " velocities: x=" + std::to_string(velocities.x) + " y=" + std::to_string(velocities.y) + " z=" + std::to_string(velocities.alt) +
                             " yaw_speed=" + std::to_string(velocities.yaw) + " successful frames: " + std::to_string(successful_frames));

            // Check if we're centered enough to consider transitioning
            float tolerance = stage == STAGE_1 ? STAGE_1_TOLERANCE : STAGE_2_TOLERANCE;
            if (errs.x < tolerance && errs.x > -1.0 * tolerance &&
                errs.y < tolerance && errs.y > -1.0 * tolerance &&
                errs.alt < tolerance && errs.alt > -1.0 * tolerance) // &&
            // errs.yaw < 5.0 && errs.yaw > -5.0)
            {
                successful_frames++;
            }
            else
            {
                successful_frames = 0;
            }

            // Transition if we've been centered for over 1 second
            if (successful_frames > 10)
            {
                success = true;
                break; // TODO: calculate FPS to determine real 1 second duration
            }

            // Dynamically adjust z velocity to only descend when close enough to being centered
            // double safe_view = 2 * (errs.alt - velocities.alt * 0.1) * tan(to_radians(CAMERA_FOV_VERTICAL / 2));
            // if (abs(errs.x) >= safe_view || abs(errs.y) >= safe_view) {
            //     velocities.alt = -0.1;
            //     velocities.yaw = 0.0;
            // }

            // Dynamic adjustment wasn't working great, but the following seemed to work better
            // Prevent the drone from descending if the target is farther than 50% of the way to the edge of the frame from the center in x or y
            float frame_size_meters_x = 640 * errs.tag_pixel_ratio;
            float frame_size_meters_y = 480 * errs.tag_pixel_ratio;
            if (abs(errs.x) >= 0.5 * (frame_size_meters_x / 2) || abs(errs.y) >= 0.5 * (frame_size_meters_y / 2))
            {
                velocities.alt = -0.1;
                velocities.yaw = 0.0;
            }

            change.forward_m_s = velocities.y;
            change.right_m_s = velocities.x;
            change.down_m_s = velocities.alt;
            change.yawspeed_deg_s = velocities.yaw;
            failed_frames = 0;
        }
        else
        {
            log(log_tag, "Failed to find Apriltag, number missed frames: " + std::to_string(failed_frames));

            if (failed_frames > 10)
            {
                change.forward_m_s = 0.0;
                change.right_m_s = 0.0;
                change.down_m_s = 0.0;
                change.yawspeed_deg_s = 0.0;
            }
            else
            {
                change.forward_m_s = velocities.y * 0.8;
                change.right_m_s = velocities.x * 0.8;
                change.down_m_s = velocities.alt * 0.8;
                change.yawspeed_deg_s = 0.0;
            }
            failed_frames++;
        }

        if (m_down > 5.0)
        {
            log(log_tag, "Emergency land, too high! Altitude: " + std::to_string(m_down), true);
            land(); // emergency land, too high
            return false;
        }

        offboard->set_velocity_body(change);
        // cv::imwrite("test_stage" + std::to_string(stage) + "_"+ std::to_string(time_span) + ".png", img); // save frame as png for debug

        high_resolution_clock::time_point t2 = high_resolution_clock::now();
        duration<double, std::milli> d = (t2 - t1);
        time_span = d.count();
        total_frames++;
    }

    if (stage == STAGE_1)
    {
        Offboard::VelocityBodyYawspeed hold{};
        offboard->set_velocity_body(hold);
    }
    else
    {
        // attempt to engage docking mechanism
        // Todo (spin servo)

        // land slowly for now instead of docking mechanism
        Offboard::VelocityBodyYawspeed hold{};
        hold.down_m_s = 0.2;
        offboard->set_velocity_body(hold);

#if PLATFORM == RASPBERRY_PI
        while (true)
        {
            if (docking_detector.is_docked())
            {
                log(log_tag, "Successfully docked!");
                for(int i = 0; i < SERVO_ATTEMPTS; i++){
                    if(docking_detector.engage_servo()){
                        break;
                        sleep_for(milliseconds(300));
                    }
                }
                break;
            }
            else
            {
                sleep_for(milliseconds(500));
            }
        }
#endif
    }

    log(log_tag, "Number of frames processed in " + std::to_string(time_span) + " ms: " + std::to_string(total_frames));
    log(log_tag, "Average FPS: " + std::to_string(total_frames / (time_span / 1000)));
    return success;
}

// Lands and disarms drone
bool Drone::land()
{
    log("Landing", "Landing");
    Offboard::Result offboard_result = offboard->stop();
    if (offboard_result != Offboard::Result::Success)
    {
        log("Landing", "Offboard stop failed, error code on next line", true);
        std::cout << "Code: " << offboard_result << std::endl;
    }

    Action::Result land_result = action->land();
    while (land_result != Action::Result::Success)
    {
        log("Landing", "Land failed, error code on next line", true);
        std::cout << "Code: " << land_result << std::endl;
        land_result = action->land();
    }

    // subscribe to telemetry updates to monitor landing progress
    const Telemetry::Result set_rate_result = telemetry->set_rate_position(1.0);
    if (set_rate_result != Telemetry::Result::Success)
    {
        log("Landing", "Telemetry Error", true);
    }
    telemetry->subscribe_position([](Telemetry::Position position)
                                  { log("Landing", "Current altitude: " + std::to_string(position.relative_altitude_m)); });
    Telemetry::LandedState curr_state = telemetry->landed_state();
    log("Landing", "Current landed state: " + get_landed_state_string(curr_state));
    telemetry->subscribe_landed_state([this, &curr_state](Telemetry::LandedState landed_state)
                                      {
                                          if (curr_state != landed_state)
                                          {
                                              curr_state = landed_state;
                                              log("Landing", "Current landed state: " + get_landed_state_string(curr_state));
                                          }
                                      });

    // wait for landing to be complete
    while (curr_state != Telemetry::LandedState::OnGround)
    {
        sleep_for(seconds(1));
    }

    // cleanup and disarm
    log("Landing", "Landed successfully " + std::to_string(telemetry->in_air()));
    Action::Result disarm_result = action->disarm();
    if (disarm_result != Action::Result::Success)
    {
        log("Landing", "Disarming failed", true);
        std::cout << "Code: " << disarm_result << std::endl;
        return false;
        ;
    }
    log("Landing", "Disarmed successfully");
    return true;
}

// Utility method to adjust errors for stage 1
void Drone::offset_errors(Errors &errs, int id)
{
    float target_offset = id <= 3 ? abs(id - 3) * 45 : (11 - id) * 45;
    float x = errs.x + BOOM_LENGTH / 2 * cos(to_radians(target_offset - errs.yaw));
    float y = errs.y + BOOM_LENGTH / 2 * sin(to_radians(target_offset - errs.yaw));
    errs.alt -= ALTITUDE_DISP;
    errs.x = x;
    errs.y = y;
    float yaw = errs.yaw + 90 - ((id - 1) * 45);
    errs.yaw = yaw;
}